lighty-core/lighty-codecs-util/src/main/java/io/lighty/codecs/util/ConverterUtils.java

```diff
@@ -161,58 +162,88 @@ public static XmlElement rpcAsOutput(final XmlElement inputXmlElement) {
         return DataSchemaContextTree.from(effectiveModelContext).findChild(yangInstanceIdentifier);
     }
 
     /**
-     * Creates an instance of {@link SchemaNode} for the given namespace, revision and localname. The
-     * namespace, revision and localname are used to construct the {@link QName} which must exist in the
-     * {@link EffectiveModelContext}.
-     *
-     * @see ConverterUtils#getSchemaNode(EffectiveModelContext, QName)
-     * @param effectiveModelContext given effective model context
-     * @param namespace {@link QName} namespace
-     * @param revision {@link QName} revision
-     * @param localName {@link QName} localname
-     * @return instance of {@link SchemaNode}
+     * Finds the {@link DataSchemaContextNode} for the given namespace, revision and local name
+     * in {@link EffectiveModelContext}.
+     *
+     * <p>
+     * Search is performed only on first level nodes of the modules, for recursive search,
+     * the {@link YangInstanceIdentifier} is needed, thus consider using
+     * {@link ConverterUtils#getSchemaNode(EffectiveModelContext, YangInstanceIdentifier)}.
+     *
+     * @param effectiveModelContext model context to search
+     * @param namespace             {@link QName} module namespace of the node to search for
+     * @param revision              {@link QName} module revision of the node to search for
+     * @param localName             {@link QName} local name of the node to search for
+     * @return optional found {@link DataSchemaContextNode}
      */
     public static Optional<@NonNull DataSchemaContextNode<?>> getSchemaNode(
-        final EffectiveModelContext effectiveModelContext, final String namespace, final String revision,
-        final String localName) {
-        QName qname = QName.create(namespace, revision, localName);
-        return DataSchemaContextTree.from(effectiveModelContext).findChild(YangInstanceIdentifier.of(qname));
+            final EffectiveModelContext effectiveModelContext, final String namespace, final String revision,
+            final String localName) {
+        return getSchemaNode(effectiveModelContext, QName.create(namespace, revision, localName));
+    }
+
+    /**
+     * Converts provided {@link YangInstanceIdentifier} to {@link SchemaNodeIdentifier.Absolute}.
+     *
+     * @param yangInstanceIdentifier yang instance identifier
+     * @return optional of converted {@link SchemaNodeIdentifier.Absolute}, empty if provided
+     *         {@link YangInstanceIdentifier} is empty
+     */
+    public static Optional<SchemaNodeIdentifier.Absolute> toSchemaNodeIdentifier(
+            final YangInstanceIdentifier yangInstanceIdentifier) {
+        final List<QName> filtered = yangInstanceIdentifier.getPathArguments().stream()
+                .filter(pa -> !(pa instanceof YangInstanceIdentifier.AugmentationIdentifier)
+                        && !(pa instanceof YangInstanceIdentifier.NodeIdentifierWithPredicates))
```

Comment:
AugmentationIdentifier has no QName so, AugmentationIdentifier.getNodeType() throws Ex, and we want to skip NodeIdentifierWithPredicates's QName since it's parent in YangInstanceIdentifier already contains this QName, if we didn't skip it, it would result in invalid SchemaPath(Absolute).

Reply:
Ok, I have tested this by my self, and deserialization works when path contains augmentation and predicates.

Reply:
Right, but try this out with something involving choice/case nodes. Does that work?